keps/sig-node/4639-oci-volume-source/README.md

```diff
@@ -0,0 +1,1240 @@
+    - Allowing direct mounting of OCI images introduces potential attack
+      vectors. Mitigation includes thorough security reviews and limiting access
+      to trusted registries. Limiting to OCI artifacts (non-runnable content)
+      and read-only mode will lessen the security risk.
+    - Path traversal attacks are a high risk for introducing security
+      vulnerabilities. Container Runtimes should re-use their existing
+      implementations to merge layers as well as secure join symbolic links in
+      the container storage prevent such issues.
+- **Compatibility Risks:** Existing webhooks watching for the images used by the pod using some policies will need to be updated to expect the image to be specified as a `VolumeSource`.
+- **Performance Risks:** Large images or artifacts could impact performance. Mitigation includes optimizations in the implementation and providing
+  guidance on best practices for users.
+
+## Design Details
+
+The new `VolumeSource` will be defined in the Kubernetes API, and the implementation will involve updating components (CRI, Kubelet)
+to support this source type. Key design aspects include:
+
+- API changes to introduce the new `VolumeSource` type.
+- Modifications to the Kubelet to handle mounting OCI images and artifacts.
+- Handling image pull secrets and registry authentication.
+- The regular OCI images (that are used to create container rootfs today) can
+  be setup similarly as a directory and mounted as a volume source.
+- For OCI artifacts, we want to convert and represent them as a directory with
+  files. A single file could also be nested inside a directory.
+
+The following code snippet illustrates the proposed API change:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: example-pod
+spec:
+  volumes:
+  - name: oci-volume
+    oci:
+      image: "example.com/my-image:latest"
+      pullPolicy: IfNotPresent
+  containers:
+  - name: my-container
+    image: busybox
+    volumeMounts:
+    - mountPath: /data
+      name: oci-volume
+```
+
+### Kubelet and Container Runtime Interface (CRI) support for OCI artifacts
+
+Kubelet and the Container Runtime Interface (CRI) currently handle OCI images. To support OCI artifacts,
+potential enhancements may be required:
+
+**Extended Media Type Handling in the container runtime:**
+   - Update container runtimes to recognize and handle new media types associated with OCI artifacts.
+   - Ensure that pulling and storing these artifacts is as efficient and secure as with OCI images.
+
+**Lifecycling and Garbage Collection:**
+   - Reuse the existing kubelet logic for managing the lifecycle of OCI objects.
+   - Extending the existing image garbage collection to not remove an OCI volume
+     image if a pod is still referencing it. 
+
+**Artifact-Specific Configuration:**
+   - Introduce new configuration options to handle the unique requirements of different types of OCI artifacts.
+
+**Artifacts as Subject Referrers:**
```

Comment:
It's unclear to me how this new volume type will interact with the referrers API. Is the idea to also mount artifacts that are referenced by the `image` field in the volume source? Can you add an example?

Reply:
This is mostly for future graduations and should be handled more in runtimes rather than Kubernetes. For example SBOMs can be referenced to verify OCI artifacts or images. But that's something we can discuss later on, I clarified that in the proposal.